lib/polyfill/pip.js

```diff
@@ -0,0 +1,232 @@
+/**
+ * @license
+ * Copyright 2016 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+goog.provide('shaka.polyfill.PiP');
+
+goog.require('shaka.log');
+goog.require('shaka.polyfill.register');
+
+
+/**
+ * @namespace shaka.polyfill.PiP
+ *
+ * @summary A polyfill to provide PiP support in Safari.
+ */
+
+
+/**
+ * Install the polyfill if needed.
+ */
+shaka.polyfill.PiP.install = function() {
+  if (!window.HTMLVideoElement) {
+    // Avoid errors on very old browsers.
+    return;
+  }
+
+  const proto = HTMLVideoElement.prototype;
+  if (proto.requestPictureInPicture &&
+    document.exitPictureInPicture) {
+    // No polyfill needed.
+    return;
+  }
+
+  if (!proto.webkitSupportsPresentationMode) {
+    // No polyfill available.
+    return;
+  }
+  shaka.log.debug('PiP.install');
+
+  /** @type {HTMLMediaElement} */
+  let polyfillPictureInPictureElement = null;
+
+  let polyfillEnterpictureinpicture = null;
```

Comment:
I think we may be avoiding compiler errors by leaving out types.  I appreciate all the work you've done on this, though, and I know the compiler can be a pain sometimes.  I think perhaps we should merge this PR as-is and I will work on some follow-up changes.  Would that be alright?

Reply:
For me it's ok!
Thanks for your revision, I really appreciate it and it help to another PR in coming.